src/huggingface_hub/commit_api.py

```diff
@@ -0,0 +1,444 @@
+        repo_id (``str``):
+            A namespace (user or an organization) and a repo name separated
+            by a `/`.
+        token (``str``):
+            An authentication token ( See https://huggingface.co/settings/tokens )
+        revision (``str``):
+            The git revision to upload the files to. Can be any valid git revision.
+
+    Raises:
+        :class:`requests.HTTPError`:
+            If the HF Hub API returned an error
+    """
+    endpoint = endpoint if endpoint is not None else ENDPOINT
+    headers = {"authorization": f"Bearer {token}"} if token is not None else None
+
+    payload = [preupload_payload(file) for file in files]
+    path2payload = {item["path"]: item for item in payload}
+
+    resp = requests.post(
+        f"{endpoint}/api/{repo_type}s/{repo_id}/preupload/{revision}",
+        json={"files": payload},
+        headers=headers,
+    )
+    resp.raise_for_status()
+
+    preupload_info: List[PreUploadFileResponse] = resp.json().get("files", [])
+    path2mode: Dict[str, UploadMode] = {
+        file["path"]: file["uploadMode"] for file in preupload_info
+    }
+
+    return [
+        CommitOperationAddAnnotated(
+            path_or_fileobj=file.path_or_fileobj,
+            path_in_repo=file.path_in_repo,
+            size=path2payload[file.path_in_repo]["size"],
+            sha=path2payload[file.path_in_repo]["sha"],
+            upload_mode=path2mode[file.path_in_repo],
+        )
+        for file in files
+    ]
+
+
+class CommitFilesPayloadFile(TypedDict):
+    path: str
+    encoding: Literal["base64"]
+    content: str
+
+
+class CommitFilesPayloadLfsFile(TypedDict):
+    path: str
+    algo: Literal["sha256"]
+    oid: str
+
+
+class CommitFilesPayloadDeletedFile(TypedDict):
+    path: str
+
+
+class CommitFilesPayload(TypedDict):
+    summary: str
+    description: str
+    files: List[CommitFilesPayloadFile]
+    lfsFiles: List[CommitFilesPayloadLfsFile]
+    deletedFiles: List[CommitFilesPayloadDeletedFile]
+
+
+def prepare_commit_payload(
+    additions: Iterable[CommitOperationAddAnnotated],
+    deletions: Iterable[CommitOperationDelete],
+    commit_summary: str,
+    commit_description: Optional[str] = None,
+) -> CommitFilesPayload:
+    commit_description = commit_description if commit_description is not None else ""
+
+    return {
+        "summary": commit_summary,
+        "description": commit_description,
+        "files": [
+            {
+                "path": file.path_in_repo,
+                "encoding": "base64",
+                "content": file.b64content.decode(),
+            }
+            for file in additions
+            if file.upload_mode == "regular"
+        ],
+        "lfsFiles": [
+            {"path": file.path_in_repo, "algo": "sha256", "oid": file.sha}
+            for file in additions
+            if file.upload_mode == "lfs"
+        ],
+        "deletedFiles": [{"path": deletion.path_in_repo} for deletion in deletions],
+    }
```

Comment:
off-topic, but did you read the code in `commands/lfs.py` @SBrandeis? (i wrote it but don't remember any of it 😱 would be good for you to read if you haven't already, out of curiosity – re. lfs protocol)